### ISSUE TITLE
Certain arbitrary changes to code results in UPX compression slowing execution of application to a crawl.

### ISSUE THREAD

**Issue by Cliffsky** [upx/upx#186]
## What's the problem (or question)?

Hi,

We've been using UPX to compress our application for the past couple months without incident (maybe a dozen different builds or so). On one occasion, the addition of a single log line to the application and its subsequent compression with UPX resulted in (totally unexpected) 6x slowdown of our application startup time, as well as slowdown of a similar magnitude throughout the operation of the application (though we haven't measured the latter bit). Memory usage or CPU usage of the application is in normal bounds, nothing out of the ordinary basically.  Eliminating the log line we added and rebuilding/compression the application fixed the issue at first. The problem recurred a month later, seemingly triggered by a similarly trivial modification to our code, in this case converting a std::wstring to an std::string in one place, which we subsequently log.

This behavior only appears in the compressed executable. The uncompressed executable shows no signs of slowdown or other odd behavior.

## What should have happened?

I suppose the ideal behavior would have been no slowdown of application at start-up or during the execution.

## Do you have an idea for a solution?

With respect to solution, it's such a ghost of a bug that us working on the application don't know where to begin. I'll list everything I've uncovered in the process of chasing it though.

Both lines of code that triggered this behavior in UPX were logging related. We're using log4cplus. We suspected at various points the existence of a memory bug in the log4cplus code that causes UPX to misbehave, but beyond the confirmation that suppressing (totally non-suspect) calls to the log4cplus logger in our application fixes the issue, we don't have any insight into how the two are related. They very well might not be.

Another thing is that the incidence of the bug depends on what computer we use to compile/compress our code. I have an AMD A10-6700 APU, my colleague has an AMD A12-9800. If I try to compile and compress the application, the resulting compressed executable suffers from this slowdown issue on whatever computer we try to run it. If I compile and compress the same build on my colleague's computer the resulting executable runs fine. Kinda adds to the untraceable nature of the bug.

## How can we reproduce the issue?

With respect to reproducing the issue, I'd have to clear it with my boss to share some of our commercial code or executables before I can really supply you guys with what you need to reproduce the issue. I guess by submitting this initial bug report i was hoping to first gauge the interest among the UPX contributors.  Also maybe a suggestion of where to start looking in the UPX source, or any off-the-top-of-your-heads ideas of what could be causing this, since we're at a loss of where to start.

## Please tell us details about your environment.

* UPX version used (`upx --version`): 3.94
* Host Operating System and version: Windows 10 build 15063
* Host CPU architecture:
* Target Operating System and version: Same as host
* Target CPU architecture: Same as host

**Comment by jreiser** [comment]
Windows 10 has some features of which upx was totally unaware; see issue #154, for instance.  That issue is being worked.  The best way to help your specific case is to provide a test case.  If it cannot be uploaded to github ("Attach files by ...") then I'm sure that you can find a way to send it to me privately.

**Comment by Cliffsky** [comment]
Great! I'll go the private route if you don't mind. What I can send you are the compressed and uncompressed binaries of the problem build. Let me know if you need anything else and I'll see if I can clear it!

**Comment by Cliffsky** [comment]
Is there an e-mail address I could use to send you the binaries?

**Comment by jreiser** [comment]
<tt>git log</tt>

**Comment by Cliffsky** [comment]
We also observed these additional details to the slowdown issue with our application startup after having been compressed with UPX.

When observing the process through Microsoft's Sysinternals tool ProcessExplorer, we noticed that our Upx-compressed application RunFixMeStick.exe was launching a lot of suspended children processes ot itself, and all those seem to be crashing with 0xc0000005 (access violation) exceptions. The exceptions seem to be silently handled.
Several dozens of process copies get launched while the application starts up.
We have a feeling it's these process copy starts are what causes slowdown to application startup.

Once few dozen of them start up, the UI of our application eventually shows up.

Afterwards, we notice sporadic starts of other suspended copies of RunFixMeStick durring execution, but at a lot more slower rate than during the startup.

Note that we attempted to disable the compression of any resources within application, but the app was still slow to start up, and we saw the same pattern of multiple suspended children processes being launched and faulting with "Access Violation 0xc0000005" exception.
We were able to track the exceptions through Windos Event viewer application.


Note that our application contains embedded resources such as embedded HTML and Javascript files, which are loaded into embedded InternetExplorer control. This control is an integral part of our user interface, and as part of startup the control is displayed and force-navigated to the HTML file embedded within executable resources. This embedded HTML file further loads other Javascript files also embedded as resources inside the executable.

Without UPX compression we only see single instance of RunFixMeStick.exe starting up, and the UI gets loaded and displayed very quickly.


Are these copies of RunFixMeStick.exe launched whenever the application is trying to access the embedded resources?

Is this a behavior exhibited by UPX compression/decompression routines embedded in our app executable?

Are there any ways to speed up the startup process?


See for your reference the following supporting material to these details:

* Video recording of progression of the RunFixMeStick.exe application startup: https://drive.google.com/open?id=11JOTMH5PSjKLX7EPzTXP92nB9C8UFAaM

* Screen capture of Event Viewer showing Access Violation exceptions triggered by children RunFixMeStick.exe processes: https://drive.google.com/open?id=1_jIvux5c8ZNjHEScmHcTOg6PU8QPcaMp

**Comment by jreiser** [comment]
<i>... our application contains embedded resources such as embedded HTML and Javascript files ...</i>  &nbsp;&nbsp;Please say more about these resources.  Can they be identified by name, type, etc?  Is every resource  one of these special resources?

**Comment by Mattiwatti** [comment]
Could you test the first scenario again but with `WerFault.exe` disabled? ([Here's](https://superuser.com/questions/1069578/disable-werfault-exe-on-crashes-for-specific-processes-on-windows-10/1069582) a good thread with some ways to do this either globally or only for your app.)

Not that I think `WerFault.exe` has anything to do with the access violation itself, but the behaviour in the video is clearly abnormal for a crashing application and makes it hard/impossible to debug.

You mention that the exceptions seem to be 'silently handled', but there isn't really such a thing. An unhandled exception will cause a process to terminate abnormally and either an error message or a JIT debugger prompt will be shown. However, you can come very close to making it look like a natural death by doing the following:
```C++
SetThreadErrorMode(SEM_FAILCRITICALERRORS, nullptr);
*((int*)0) = 5; // SEGVICIDE in silence
```
`WerFault.exe` does exactly this when it murders a process that has crashed after it has finished harvesting it for ~~organs~~ juicy telemetry data. The complete process is a bit more involved, as you can't normally change another thread's error mode. Instead WerFault uses ntdll's `RtlCreateProcessReflection` to fork the dead process to make a zombie clone that it can then inject a thread into to do its dirty work in peace. Except on your system it seems to be making way too many forks and failing to cover up the bloodbath.

This is just my hypothesis, but IMO it's a bit more likely than your process crashing, resurrecting itself, randomly multiplying by 20 and then suspending itself again.

**Comment by Cliffsky** [comment]
Hey Mattiwatti,

Our application doesn't actually crash - the instance we manually launch persists the entire duration of these other processes being launched. Also nowhere in our code does the application try to launch itself - either as a response to crashing or otherwise. So what we see is  more like our application continually and inexplicably launching copies of itself - we're assuming because it's trying to access a compressed resource but we're not sure. 

We're going to look closer at the behavior UPX exhibits compressing the same build across different machines. Again we're seeing that the incidence of the bug depends on which machine compiled the executable to be compressed. We're suspecting differences in the Visual Studio configuration might be at play here. 

With respect to the resources, we've got a dozen or so CSS files, three eot font files, our icon, a dozen split up javascript files, one html file and some pngs. That said we've seen the case where specifying UPX not to compress any of the resources nevertheless leads to the buggy (i.e. slow) behavior. We're not sure if that precludes our above assumption (that these child processes are being spawned as part of an attempt by our application to access resources).

Anyway we'll keep posting as we keep digging. Thanks for continuing to give this issue attention!

**Comment by Mattiwatti** [comment]
Actually, none of what you say is inconsistent with process cloning :) Here's my attempt at recreating the process tree from your video (albeit without the nice looking icons...) using only a single `CreateProcess` call to create the parent process. All child processes and their address spaces are cloned from this process, and (importantly), no section object was created for any child processes, as they simply inherit the parent's section and increase the reference count of the object.

![clones](https://i.imgur.com/xW6Uhtz.png)

Of course, I cheated here and the parent process is actually cloning itself. But note that this can be done by any process with an open handle to another process so long as the handle has sufficient access rights.

Whether the process is healthy or on the way out due to a crash is irrelevant - as long as you can obtain a handle to a process, it can be cloned. Of course, logic dictates that the error reporting guy only comes in when there's been an error, so I don't have an explanation for that part of the puzzle. I did however find another code path that could cause a process to be cloned: `ntdll!PssNtCaptureSnapshot` can do this and will eventually call `NtCreateProcessEx`, which (despite its name) will clone the specified process rather than create one, due to being given a NULL section handle and a parent process to clone the address space from. This function is only ever called by the unhandled exception filter.

It's not possible that either UPX or the ntdll loader are creating processes due to something as mundane as failing to find a resource or something like that. UPX doesn't even _have_ any code related to process creation, so that makes that part simple. The ntdll loader technically _could_ create processes if it wanted to, but apart from that being pointless here, creating a process is such a ridiculously costly operation performance-wise compared to anything else the loader does that this would never be considered a serious option. The ntdll loader _loves_ killing your process and would never pass up a good excuse to do so.

Like @jreiser, I'm also willing to take a look at this if it's possible for you to send me a file to reproduce the issue (source code not necessary). The working uncompressed file along with the slow/non-working compressed output and the command line used to generate it should suffice. Contact is <_my_github_username_>@gmail.com.

**Comment by Cliffsky** [comment]
Gotcha!

**Comment by Mattiwatti** [comment]
OK, I received a sample from @Cliffsky and took a look at it to investigate. It took some time before I could reproduce the issue, because it requires a specific combination of maladies to be present on the machine:
- Windows 10
- Windows Error Reporting installed and enabled

Of course, being a die hard hater of both, and loving to see my prejudices confirmed, I secretly already expected this. The problem is that getting Windows Error Reporting back and working is really hard if you've slipstreamed it out of the installation media, disabled it via both the system settings and group policy, and deleted the service from the registry... like I apparently did on all 5 of my Windows 10 VMs. So, one Windows 10 installation from the official MS media later, I finally managed to reproduce the issue. It is indeed caused by `WerFault.exe` cloning the process, specifically via the second method I mentioned (`ntdll!PssNtCaptureSnapshot`):
```
00 c0f33be8 8176c50e 0038fb88 02000000 00000000 nt!NtCreateProcessEx
01 c0f33be8 773a16f0 0038fb88 02000000 00000000 nt!KiSystemServicePostCall
02 0038fb38 773a038a 77331a23 0038fb88 02000000 ntdll!KiFastSystemCallRet
03 0038fb3c 77331a23 0038fb88 02000000 00000000 ntdll!NtCreateProcessEx+0xa
04 0038fbc8 746108ab 0038fc08 0000013c d00039ff ntdll!PssNtCaptureSnapshot+0x2c3
05 0038fbe0 009e8662 0000013c d00039ff 0001003f KERNELBASE!PssCaptureSnapshot+0x1b
06 0038fd08 009e8949 00000008 02bb0f76 00000001 WerFault!CaptureSnapshot+0x185
07 0038fd30 009dbb0d 00000000 00000001 00000000 WerFault!SnapshotCaptureMain+0x248
08 0038fd54 009f8e5a 00000008 02bb0f08 02bb28e8 WerFault!wmain+0x290
09 0038fd98 75259564 00514000 75259540 34621d7b WerFault!operator delete+0x211
0a 0038fdac 773729ac 00514000 826e18fc 00000000 KERNEL32!BaseThreadInitThunk+0x24
0b 0038fdf4 77372980 ffffffff 773b3cd7 00000000 ntdll!__RtlUserThreadStart+0x2b
0c 0038fe04 00000000 009f8f10 00514000 00000000 ntdll!_RtlUserThreadStart+0x1b
```

Note that this is only one of at least two `WerFault.exe`s launched - _per thread of the executable_. There's one to harvest your private data, and another to upload it to Microsoft. The exact behaviour depends on the command line it is launched with. (I'll skip the stacktrace for the other one since it is about 4 times as long and not very interesting.) Because of the unfortunate choice of using a self-balancing _thread pool with worker threads_ to parallelize the ntdll loader, you can imagine that this spirals out of control pretty quickly, at least until a throttle limit on either the error reporting service or the thread pool puts a stop to it and no new processes spawn.

OK, so that's **what's** happening - but **why** is it happening? The answer can be found in the worker threads that are happily spinning up to inevitably end up in the same place as the previous one:
```
0, ntoskrnl.exe!KiCommitThreadWait+0x17e
1, ntoskrnl.exe!KeWaitForMultipleObjects+0x1f4
2, ntoskrnl.exe!ObWaitForMultipleObjects+0x29f
3, ntoskrnl.exe!NtWaitForMultipleObjects+0xfd
4, ntoskrnl.exe!KiSystemServicePostCall
5, ntdll.dll!KiFastSystemCallRet
6, ntdll.dll!NtWaitForMultipleObjects+0xa
7, ntdll.dll!WerpWaitForCrashReporting+0x88
8, ntdll.dll!RtlReportExceptionHelper+0x226
9, ntdll.dll!RtlReportException+0x61
10, ntdll.dll!LdrpCalloutExceptionFilter+0x35
11, ntdll.dll!LdrpCallTlsInitializers+0x60ec4 (No unwind info)
12, ntdll.dll!_EH4_CallFilterFunc+0x12
13, ntdll.dll!_except_handler4_common+0x7f
14, ntdll.dll!_except_handler4+0x20
15, ntdll.dll!ExecuteHandler2+0x26 (No unwind info)
16, ntdll.dll!ExecuteHandler+0x24 (No unwind info)
17, ntdll.dll!KiUserExceptionDispatcher+0x26
18, 0x9e7ef80 (No unwind info)
19, 0x9e7ef9c (No unwind info)
```
**LdrpCallTlsInitializers** in frame no. 11 gives away the underlying cause (despite the weird offset the symbol name is correct, this is the function's exception handler). So, there's something wrong with TLS, but what? Let's check out the working uncompressed version:
![uncompr](https://i.imgur.com/29vuSiI.png)
If you don't know the meaning of these fields, that's good and I would recommend keeping it that way for your sanity. But they are described in section 6.7 of the [PE/COFF specification](http://www.osdever.net/documents/PECOFF.pdf). The short version is that every address in the TLS directory is a VA (in this case, all in **.rdata**), except for `AddressOfIndex` which is an RVA (in this case, in **.data**). Additionally, there is no size field for the callbacks; rather it is a null-terminated array of pointers, much the same way a C string is null-terminated array of chars.

Now let's compare this to the compressed version:
![compr](https://i.imgur.com/cKRPKDD.png)
Can you tell which fields are wrong? (if you look closely you can see I've added subtle hints to the picture.)

The array of callbacks being incorrect is a pretty safe bet because of the changed size alone, and the clearly invalid `0x98` pointer at the end. `AddressOfIndex` on the other hand seems to be another case where the loader has started complaining about VAs that lie within virtual sections (i.e. UPX0). I guessed that this might be the case because the load config directory gave the same issues with the Windows 8.1/10 loader (see #154). I did some hex editing to test this hypothesis and ended up with this:
![fix_kinda](https://i.imgur.com/nWwNm4A.png)

And this does indeed work! However, note that with this 'fix' I've made the `AddressOfIndex` simply point to some unused space in UPX1. In other words, a PE image won't receive its proper TLS index with this hack because it is written to a different location than it is expecting. However, the loader normally assigns the main exe in a process TLS index 0(\*). Since the default value is also 0, this happens to work even if the field is never written to. DLLs are another story: since they will receive TLS indices >= 1, using index 0 will cause them to either overwrite the TLS data of the exe or the other way around.
(\*) There is no law that says this must be true, but I've never seen any exceptions to this.

It's possible that this is actually the same issue mr. 740 MB DLL guy is having (#188), but that the loader doesn't tolerate the TLS directory in his case because the file is a DLL, whereas an exe is allowed to continue because it would have received index 0 anyway. That's just speculation though.

Hopefully the info above is useful to @jreiser, because the UPX codebase is way over my head and I won't be able to fix it myself.

Lastly, I was lying when I said Windows 10 is required to reproduce this. It actually works in Windows 8.1 too! The difference is that in 8.1, only 1 `WerFault.exe` instance is launched, which immediately exits and does not leave 20 suspended zombie child processes behind like on Windows 10. So why won't the Windows 10 ones die? Well, the cloned processes are fairly unique in that they have **0 threads**. A process is normally terminated when its thread count reaches 0, except for processes that were _created_ with 0 threads (this is only possible with `NtCreateProcessEx`, not the newer `NtCreateUserProcess`). Unless you create a thread in such a process, it dies when the last handle to it is closed. So, if all `WerFault.exe`s are gone, who is keeping open handles to these processes around? It's your own process! Of course, the handles were forcefully shoved in there by `WerFault.exe` via `NtDuplicateObject` in the first place, and then simply never closed. Methinks there may be another bug in here that isn't related to UPX at all...

**Comment by ml1050** [comment]
Thanks @Mattiwatti for looking at this. 

It seems to me that the main problem here is that the TLS callbacks area should only contain 1 entry (which uncompresses the image). It looks like the terminating zero is missing. Can you please check whether the following patch fixes this?

```
diff --git a/src/pefile.cpp b/src/pefile.cpp
index 8146fb6..2e72e6e 100644
--- a/src/pefile.cpp
+++ b/src/pefile.cpp
@@ -1404,6 +1404,7 @@ void PeFile::processTls2(Reloc *rel,const Interval *iv,unsigned newaddr,
     {
       //set handler offset
       *(LEXX*)(otls + sotls - 2 * cb_size) = tls_handler_offset + imagebase;
+      *(LEXX*)(otls + sotls - cb_size) = 0;
       //add relocation for TLS handler offset
       rel->add(newaddr + sotls - 2 * cb_size, reloc_type);
     }
```

UPX does not move the TLS index to another address by design. The main reason for it is that exe files with no relocation information would not work that way, as UPX can not find the references to that TLS index to relocate them. This means some hacks are needed when the TLS index points to the compressed data section. This is implemented in the PETLSHAK assembly section.

**Comment by Mattiwatti** [comment]
@ml1050 I tested the patch and it works precisely as you'd expect. The callback array is now correctly terminated. Nice one!

What's driving me crazy is that I compressed the file (using only your patch without moving `AddressOfIndex`) and now it suddenly just works. Even if I compress using the unpatched version and manually terminate the array - no more issues. I'm pretty sure I terminated the array long before I even considered changing `AddressOfIndex` though, as the callback array was the clearly incorrect one. Only when that still gave me the `WerFault.exe` avalanche did I move the TLS index. Weird...

Good point re: non-relocatable images, I hadn't considered that. Won't that still cause problems elsewhere with the Windows 8.1+ loader though? I'm thinking mainly of the security cookie VA and the CFG function pointers in the load configuration directory.

**Comment by ml1050** [comment]
Thanks @Mattiwatti for testing the patch. The final patch will be different, because I've just found the bug which copied some garbage to the previously zeroed area of the callbacks.

About the load config directory. Unfortunately there is no real support in UPX for most of the "latest" windows loader features. The last windows version I used was W95, and at that time I debugged the kernel using winice :-) Since then I use linux exclusively, and I was only able to fix bugs in UPX which I could reproduce under wine. New features like 64 bit support and TLS callbacks were contributed by a non core developer.
